File: common/src/main/java/com/ishow/common/widget/viewpager/looping/lndicator/DefaultLoopingIndicator.java
package com.ishow.common.widget.viewpager.looping.lndicator;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import com.ishow.common.utils.UnitUtils;
import com.ishow.common.widget.viewpager.looping.ILoopingIndicator;

/**
 * Created by yuhaiyang on 2018/9/13.
 * 默认的
 */
public class DefaultLoopingIndicator implements ILoopingIndicator {
    private Paint mIndicatorPaint;
    private Paint mIndicatorBorderPaint;
    private int mIndicatorRadius;
    private int mIndicatorItemWidth;
    private int mIndicatorWidth;
    private int mIndicatorHeight;

    private int mWidth;
    private int mHeight;

    public DefaultLoopingIndicator() {
        mIndicatorRadius = UnitUtils.dip2px(3);
        mIndicatorItemWidth = 5 * mIndicatorRadius;
        mIndicatorHeight = 6 * mIndicatorRadius;

        mIndicatorPaint = new Paint();
        mIndicatorPaint.setAntiAlias(true);
        mIndicatorPaint.setDither(true);
        mIndicatorPaint.setColor(Color.WHITE);

        mIndicatorBorderPaint = new Paint();
        mIndicatorBorderPaint.setAntiAlias(true);
        mIndicatorBorderPaint.setDither(true);
        mIndicatorBorderPaint.setColor(Color.GRAY);
        mIndicatorBorderPaint.setStyle(Paint.Style.STROKE);
    }

    @Override
    public void onViewSizeChanged(int w, int h) {
        mWidth = w;
        mHeight = h;
    }

    @Override
    public void onDataSizeChanged(int count) {
        mIndicatorWidth = mIndicatorItemWidth * count;
    }

    @Override
    public void onDraw(Canvas canvas, final int scrollX, final int count, final int currentPosition, final float positionOffset) {
        final int nextPosition = currentPosition == count - 1 ? 0 : currentPosition + 1;
        final int startX = mWidth / 2 - mIndicatorWidth / 2 + scrollX;
        final int y = mHeight - mIndicatorHeight;
        int currentIndicatorRadius = (int) (mIndicatorRadius + mIndicatorRadius * (1 - positionOffset) * 0.38F);
        int nextIndicatorRadius = (int) (mIndicatorRadius + mIndicatorRadius * positionOffset * 0.38F);
        for (int i = 0; i < count; i++) {
            int x = startX + i * mIndicatorItemWidth + (mIndicatorItemWidth - mIndicatorRadius) / 2;
            if (i == currentPosition) {
                canvas.drawCircle(x, y, currentIndicatorRadius, mIndicatorPaint);
                canvas.drawCircle(x, y, currentIndicatorRadius, mIndicatorBorderPaint);
            } else if (i == nextPosition) {
                canvas.drawCircle(x, y, nextIndicatorRadius, mIndicatorPaint);
                canvas.drawCircle(x, y, nextIndicatorRadius, mIndicatorBorderPaint);
            } else {
                canvas.drawCircle(x, y, mIndicatorRadius, mIndicatorPaint);
                canvas.drawCircle(x, y, mIndicatorRadius, mIndicatorBorderPaint);
            }
        }
    }
}

File: common/src/main/java/com/ishow/common/widget/viewpager/looping/indicator/DefaultLoopingIndicator.kt
package com.ishow.common.widget.viewpager.looping.indicator

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint

import com.ishow.common.utils.UnitUtils
import com.ishow.common.widget.viewpager.looping.ILoopingIndicator

/**
 * Created by yuhaiyang on 2018/9/13.
 * 默认的
 */
class DefaultLoopingIndicator : ILoopingIndicator {
    private val mIndicatorPaint: Paint
    private val mIndicatorBorderPaint: Paint
    private val mIndicatorRadius: Int = UnitUtils.dip2px(3)
    private val mIndicatorItemWidth: Int
    private var mIndicatorWidth: Int = 0
    private val mIndicatorHeight: Int

    private var mWidth: Int = 0
    private var mHeight: Int = 0

    init {
        mIndicatorItemWidth = 5 * mIndicatorRadius
        mIndicatorHeight = 6 * mIndicatorRadius

        mIndicatorPaint = Paint()
        mIndicatorPaint.isAntiAlias = true
        mIndicatorPaint.isDither = true
        mIndicatorPaint.color = Color.WHITE

        mIndicatorBorderPaint = Paint()
        mIndicatorBorderPaint.isAntiAlias = true
        mIndicatorBorderPaint.isDither = true
        mIndicatorBorderPaint.color = Color.GRAY
        mIndicatorBorderPaint.style = Paint.Style.STROKE
    }

    override fun onViewSizeChanged(w: Int, h: Int) {
        mWidth = w
        mHeight = h
    }

    override fun onDataSizeChanged(count: Int) {
        mIndicatorWidth = mIndicatorItemWidth * count
    }

    override fun onDraw(canvas: Canvas, scrollX: Int, count: Int, currentPosition: Int, positionOffset: Float) {
        val nextPosition = if (currentPosition == count - 1) 0 else currentPosition + 1
        val startX = mWidth / 2 - mIndicatorWidth / 2 + scrollX
        val y = mHeight - mIndicatorHeight
        val currentIndicatorRadius = (mIndicatorRadius + mIndicatorRadius.toFloat() * (1 - positionOffset) * 0.38f).toInt()
        val nextIndicatorRadius = (mIndicatorRadius + mIndicatorRadius.toFloat() * positionOffset * 0.38f).toInt()
        for (i in 0 until count) {
            val x = startX + i * mIndicatorItemWidth + (mIndicatorItemWidth - mIndicatorRadius) / 2
            when (i) {
                currentPosition -> {
                    canvas.drawCircle(x.toFloat(), y.toFloat(), currentIndicatorRadius.toFloat(), mIndicatorPaint)
                    canvas.drawCircle(x.toFloat(), y.toFloat(), currentIndicatorRadius.toFloat(), mIndicatorBorderPaint)
                }
                nextPosition -> {
                    canvas.drawCircle(x.toFloat(), y.toFloat(), nextIndicatorRadius.toFloat(), mIndicatorPaint)
                    canvas.drawCircle(x.toFloat(), y.toFloat(), nextIndicatorRadius.toFloat(), mIndicatorBorderPaint)
                }
                else -> {
                    canvas.drawCircle(x.toFloat(), y.toFloat(), mIndicatorRadius.toFloat(), mIndicatorPaint)
                    canvas.drawCircle(x.toFloat(), y.toFloat(), mIndicatorRadius.toFloat(), mIndicatorBorderPaint)
                }
            }
        }
    }
}
